python/ray/experimental/state/api.py

```diff
@@ -43,37 +45,96 @@
 
     def __init__(
         self,
-        api_server_address: str = None,
+        address: Optional[str] = None,
+        cookies: Optional[Dict[str, Any]] = None,
+        headers: Optional[Dict[str, Any]] = None,
     ):
+        """Initialize a StateApiClient and check the connection to the cluster.
+
+        Args:
+            address: The address of Ray API server. If not provided,
+                it will be configured automatically from querying the GCS server.
+            cookies: Cookies to use when sending requests to the HTTP job server.
+            headers: Headers to use when sending requests to the HTTP job server, used
+                for cases like authentication to a remote cluster.
+        """
+        if requests is None:
+            raise RuntimeError(
+                "The Ray state CLI & SDK require the ray[default] "
+                "installation: `pip install 'ray[default']``"
+            )
+        if not headers:
+            headers = {"Content-Type": "application/json"}
         super().__init__(
-            self._get_default_api_server_address()
-            if api_server_address is None
-            else api_server_address,
+            address,
             create_cluster_if_needed=False,
-            headers={"Content-Type": "application/json"},
+            headers=headers,
+            cookies=cookies,
         )
 
-    @classmethod
-    def _get_default_api_server_address(cls) -> str:
-        assert (
-            ray.is_initialized()
-            and ray._private.worker.global_worker.node.address_info["webui_url"]
-            is not None
-        )
-        return (
-            f"http://{ray._private.worker.global_worker.node.address_info['webui_url']}"
+    def print_slow_api_warning(
+        self,
+        timeout: int,
+        cv: threading.Condition,
+        endpoint: str,
+        print_interval_s: float,
+        print_warning: bool,
+    ):
+        """Print the slow API warning. every print_interval_s seconds until timeout.
+
+        Args:
+            timeout: The API timeout.
+            cv: Condition variable. If it is notified, the thread will terminate.
+            endpoint: The endpoint to query.
+            print_interval_s: The interval to print warning message until timeout.
+            print_warning: Print the warning if it is set to True.
+        """
+        start = time.time()
+        last = time.time()
+        i = 1
+        while True:
+            with cv:
+                # Will return True if it is notified from other thread
+                # which means it should terminate the loop.
+                # False if timeout is occured.
+                notified_to_terminate = cv.wait(0.1)
+                if notified_to_terminate:
+                    break
+
+            elapsed = time.time() - start
+            if time.time() - last >= print_interval_s:
+                if print_warning:
+                    print(
```

Comment:
can we print it to stderr instead of stdout? otherwise it hurts inter-operatability with other bash tools.